File: frontend/src/api/leads.ts
import { http } from '@/lib/http';

import {
  type Anexo,
  AnexoSchema,
  type Lead,
  LeadSchema,
  type LeadPrioridade,
  type LeadStatus,
  type Observacao,
  ObservacaoSchema,
} from './schemas';
import { z } from 'zod';

export interface LeadFilters {
  stage_id?: string;
  representante_id?: string;
  cliente_id?: string;
  status?: LeadStatus;
  prioridade?: LeadPrioridade;
  busca?: string;
  incluir_excluidos?: boolean;
}

function toQuery(f: LeadFilters): string {
  const params = new URLSearchParams();
  Object.entries(f).forEach(([k, v]) => {
    if (v !== undefined && v !== null && v !== '') params.set(k, String(v));
  });
  const qs = params.toString();
  return qs ? `?${qs}` : '';
}

export async function listLeads(filters: LeadFilters = {}): Promise<Lead[]> {
  const data = await http<unknown>(`/api/v1/leads${toQuery(filters)}`);
  return z.array(LeadSchema).parse(data);
}

export async function getLead(id: string): Promise<Lead> {
  const data = await http<unknown>(`/api/v1/leads/${id}`);
  return LeadSchema.parse(data);
}

export interface LeadCreatePayload {
  cliente_id: string;
  representante_id?: string;
  stage_id: string;
  projeto?: string;
  valor?: number;
  prioridade?: LeadPrioridade;
  tags?: string[];
  forma_pagamento?: string;
  valor_entrada?: number;
  percentual_entrada?: number;
}

export async function createLead(payload: LeadCreatePayload): Promise<Lead> {
  const data = await http<unknown>('/api/v1/leads', { method: 'POST', body: payload });
  return LeadSchema.parse(data);
}

export interface LeadUpdatePayload {
  projeto?: string | null;
  valor?: number;
  prioridade?: LeadPrioridade;
  tags?: string[];
  data_ultimo_contato?: string | null;
  tipo_ultimo_contato?: string | null;
  projeto_2d_enviado?: boolean;
  projeto_2d_data?: string | null;
  projeto_3d_enviado?: boolean;
  projeto_3d_data?: string | null;
  probabilidade_override?: number | null;
  valor_entrada?: number | null;
  percentual_entrada?: number | null;
  forma_pagamento?: string | null;
}

export async function updateLead(leadId: string, payload: LeadUpdatePayload): Promise<Lead> {
  const data = await http<unknown>(`/api/v1/leads/${leadId}`, { method: 'PATCH', body: payload });
  return LeadSchema.parse(data);
}

export async function moveLeadStage(leadId: string, stage_id: string): Promise<Lead> {
  const data = await http<unknown>(`/api/v1/leads/${leadId}/move-stage`, {
    method: 'POST',
    body: { stage_id },
  });
  return LeadSchema.parse(data);
}

export async function concluirLead(
  leadId: string,
  resultado: 'ganho' | 'perdido',
  motivo_perda?: string,
): Promise<Lead> {
  const data = await http<unknown>(`/api/v1/leads/${leadId}/concluir`, {
    method: 'POST',
    body: { resultado, motivo_perda },
  });
  return LeadSchema.parse(data);
}

export async function deleteLead(leadId: string): Promise<void> {
  await http<void>(`/api/v1/leads/${leadId}`, { method: 'DELETE' });
}

// ── Observações ─────────────────────────────────────────────────────
export async function listObservacoes(leadId: string): Promise<Observacao[]> {
  const data = await http<unknown>(`/api/v1/leads/${leadId}/observacoes`);
  return z.array(ObservacaoSchema).parse(data);
}

export async function addObservacao(leadId: string, texto: string): Promise<Observacao> {
  const data = await http<unknown>(`/api/v1/leads/${leadId}/observacoes`, {
    method: 'POST',
    body: { texto, tipo: 'manual' },
  });
  return ObservacaoSchema.parse(data);
}

// ── Anexos ─────────────────────────────────────────────────────────
export async function listAnexos(leadId: string): Promise<Anexo[]> {
  const data = await http<unknown>(`/api/v1/anexos/lead/${leadId}`);
  return z.array(AnexoSchema).parse(data);
}

export async function uploadAnexo(leadId: string, file: File): Promise<Anexo> {
  const form = new FormData();
  form.append('file', file);
  const data = await http<unknown>(`/api/v1/anexos/lead/${leadId}`, {
    method: 'POST',
    body: form,
  });
  return AnexoSchema.parse(data);
}

export async function deleteAnexo(anexoId: string): Promise<void> {
  await http<void>(`/api/v1/anexos/${anexoId}`, { method: 'DELETE' });
}

export function getAnexoDownloadUrl(anexoId: string): string {
  return `/api/v1/anexos/${anexoId}/download`;
}

File: backend/app/api/v1/leads.py
"""Endpoints de Leads."""

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.deps.auth import CurrentUserDep, get_current_user
from app.deps.db import get_session
from app.models.lead import LeadPrioridade, LeadStatus
from app.repositories.leads import LeadRepository
from app.repositories.observacoes import ObservacaoRepository
from app.schemas.lead import (
    LeadConcluir,
    LeadCreate,
    LeadFilters,
    LeadMoveStage,
    LeadOut,
    LeadUpdate,
)
from app.schemas.observacao import ObservacaoCreate, ObservacaoOut
from app.services.leads import (
    InvalidTransition,
    LeadNotFound,
    LeadService,
    StageNotFound,
)

router = APIRouter(prefix="/leads", tags=["leads"], dependencies=[Depends(get_current_user)])


@router.get("", response_model=list[LeadOut])
async def list_leads(
    _user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
    stage_id: UUID | None = Query(default=None),
    representante_id: UUID | None = Query(default=None),
    cliente_id: UUID | None = Query(default=None),
    status_filter: LeadStatus | None = Query(default=None, alias="status"),
    prioridade: LeadPrioridade | None = Query(default=None),
    incluir_excluidos: bool = Query(default=False),
    busca: str | None = Query(default=None, max_length=200),
    limit: int = Query(default=500, ge=1, le=1000),
    offset: int = Query(default=0, ge=0),
) -> list[LeadOut]:
    filters = LeadFilters(
        stage_id=stage_id,
        representante_id=representante_id,
        cliente_id=cliente_id,
        status=status_filter,
        prioridade=prioridade,
        incluir_excluidos=incluir_excluidos,
        busca=busca,
    )
    repo = LeadRepository(session)
    items = await repo.list_all(filters, limit=limit, offset=offset)
    return [LeadOut.model_validate(l) for l in items]


@router.post("", response_model=LeadOut, status_code=status.HTTP_201_CREATED)
async def create_lead(
    payload: LeadCreate,
    user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> LeadOut:
    service = LeadService(session)
    try:
        lead = await service.create(
            tenant_id=user.tenant_id,
            criado_por=user.id,
            autor_nome=user.nome,
            payload=payload,
        )
    except StageNotFound as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e
    return LeadOut.model_validate(lead)


@router.get("/{lead_id}", response_model=LeadOut)
async def get_lead(
    lead_id: UUID,
    _user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> LeadOut:
    repo = LeadRepository(session)
    lead = await repo.get_by_id(lead_id)
    if lead is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND)
    return LeadOut.model_validate(lead)


@router.patch("/{lead_id}", response_model=LeadOut)
async def update_lead(
    lead_id: UUID,
    payload: LeadUpdate,
    _user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> LeadOut:
    repo = LeadRepository(session)
    lead = await repo.get_by_id(lead_id)
    if lead is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND)
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(lead, key, value)
    from app.db.base import utcnow

    lead.data_ultima_movimentacao = utcnow()
    await session.flush()
    return LeadOut.model_validate(lead)


@router.delete("/{lead_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_lead(
    lead_id: UUID,
    _user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> None:
    repo = LeadRepository(session)
    lead = await repo.get_by_id(lead_id)
    if lead is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND)
    await repo.soft_delete(lead)


@router.post("/{lead_id}/move-stage", response_model=LeadOut)
async def move_stage(
    lead_id: UUID,
    payload: LeadMoveStage,
    user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> LeadOut:
    service = LeadService(session)
    try:
        lead = await service.move_stage(
            tenant_id=user.tenant_id,
            lead_id=lead_id,
            new_stage_id=payload.stage_id,
            autor_id=user.id,
            autor_nome=user.nome,
        )
    except LeadNotFound as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e)) from e
    except StageNotFound as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e
    except InvalidTransition as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e)) from e
    return LeadOut.model_validate(lead)


@router.post("/{lead_id}/concluir", response_model=LeadOut)
async def concluir_lead(
    lead_id: UUID,
    payload: LeadConcluir,
    user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> LeadOut:
    service = LeadService(session)
    try:
        lead = await service.concluir(
            tenant_id=user.tenant_id,
            lead_id=lead_id,
            resultado=payload.resultado,
            motivo_perda=payload.motivo_perda,
            autor_id=user.id,
            autor_nome=user.nome,
        )
    except LeadNotFound as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e)) from e
    except InvalidTransition as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e)) from e
    return LeadOut.model_validate(lead)


# ── Observações do lead ────────────────────────────────────────────
@router.get("/{lead_id}/observacoes", response_model=list[ObservacaoOut])
async def list_observacoes(
    lead_id: UUID,
    _user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[ObservacaoOut]:
    repo = ObservacaoRepository(session)
    items = await repo.list_by_lead(lead_id)
    return [ObservacaoOut.model_validate(o) for o in items]


@router.post(
    "/{lead_id}/observacoes",
    response_model=ObservacaoOut,
    status_code=status.HTTP_201_CREATED,
)
async def add_observacao(
    lead_id: UUID,
    payload: ObservacaoCreate,
    user: CurrentUserDep,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> ObservacaoOut:
    from uuid import uuid4

    from app.models.observacao import Observacao

    repo_lead = LeadRepository(session)
    if (await repo_lead.get_by_id(lead_id)) is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="lead não encontrado")

    obs = Observacao(
        id=uuid4(),
        tenant_id=user.tenant_id,
        lead_id=lead_id,
        autor_id=user.id,
        autor_nome=user.nome,
        texto=payload.texto,
        tipo=payload.tipo,
    )
    repo = ObservacaoRepository(session)
    await repo.add(obs)
    return ObservacaoOut.model_validate(obs)
